# Objective
Get hands-on experience with Transformer models by generating text using a pretrained model like GPT-2. This project introduces you to the core ideas behind Transformer-based architectures and helps you explore how these models generate coherent, contextually relevant text.

# Project Details

You will use a pretrained Transformer model (e.g., GPT-2) within a Jupyter Notebook environment to:

- Generate text based on different prompts.

- Experiment with generation parameters.


Reflect on how Transformers understand and extend language sequences.s.uences.

In [3]:
!pip install transformers torch
!pip install huggingface_hub[hf_xet]

In [4]:
import warnings
warnings.filterwarnings("ignore")

In [5]:
from transformers import pipeline

# Load GPT-2 text generation model
generator = pipeline("text-generation", model="gpt2")

# Set your prompt
prompt = "In the future, education will"

# Generate text
result = generator(prompt,  max_new_tokens=100, temperature=0.7)
print(result[0]['generated_text'])

Device set to use cpu
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


In the future, education will be used to improve the quality of life of Canadians, to build and maintain a strong economy, to promote healthy living, and to promote responsible transportation.

We have a responsibility to make sure that Canadians have the tools that will enable them to become self-sufficient, and to ensure that the quality of life of our country is not just increased, but increased for the better.

To further this goal, we're offering a new program called the Canadian Job Creation Program. As part of the


In [6]:
# Experiment with different prompts
prompt = 'The impact of AI on the future of work'
result = generator(prompt, max_new_tokens=100, temperature=0.8)
print(result[0]['generated_text'])

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


The impact of AI on the future of work is very real.

A lot of researchers are thinking about what the future might look like. The first big research question was how would AI drive innovation and innovation in the field of medicine? Well, this has gone around and it was always that way. As long as you had that type of data that you wanted, the problem of how to make progress is basically gone.

[Laughs.]

I think it's safe to say that the most important thing that we don't want


In [7]:
prompt = 'Once upon a time, there was a kingdom'
result = generator(prompt,  max_new_tokens=200, temperature=0.6)
print(result[0]['generated_text'])

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Once upon a time, there was a kingdom of the heavens, and a throne of the earth; and they were like unto the king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth; and they were like unto the king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth, and a king of the earth


# Reflection

Transformers are a type of neural network architecture that has revolutionized the field of artificial intelligence, particularly in natural language processing (NLP). Unlike traditional models such as RNNs or LSTMs, transformers use a mechanism called self-attention to weigh the importance of each word in a sentence, allowing them to capture long-range dependencies more effectively. This architecture is the foundation for many modern language models, including GPT-2, BERT, and GPT-3. Their ability to handle vast amounts of text and generate coherent responses makes them essential tools in AI today.

For this experiment, I used the GPT-2 model provided by Hugging Face’s transformers library along with the pipeline API for text generation. GPT-2 is a large language model pre-trained on internet text and fine-tuned for various downstream tasks. I tested the model in a Python environment and experimented with different prompts and temperature settings to see how the output changed.

The prompts I tried included "In the future, education will", "The impact of AI on the future of work", "Once upon a time, there was a kingdom". Each prompt was tested with variations in temperature (e.g., 0.6, 0.7, 0.8) and max_new_tokens (e.g., 100, 200) to observe how the model’s creativity and coherence changed.

With the prompt "In the future, education will" and a temperature of 0.7, the output was relatively formal and focused on educational reform and inclusivity. It stayed on-topic and mentioned that education will be used to improve the quality of life of Canadians.

For "The impact of AI on the future of work" with a temperature of 0.8, the response became more expansive and opinionated. It included more abstract thoughts about how AI would drive innovation. The higher temperature led to slightly more diverse and creative text.

The prompt "Once upon a time, there was a kingdom" with a max_new_tokens of 200 and temperature 0.6 generated a repetitive narrative. It repeated phrases like "a king of the earth," which indicated a limitation in the model’s output control and coherence for longer generations.

Adjusting the temperature clearly influenced how creative or predictable the model was. Lower values led to safer, more structured responses, while higher values introduced more surprising or imaginative continuations. Changing the prompt type also affected the domain and tone of the output. Formal prompts generated formal writing, while story-like prompts led to narrative or fairy tale-style continuations.

This experiment highlighted the powerful capabilities of transformer-based models like GPT-2 in generating coherent and contextually relevant text from a simple prompt. I learned how even small changes in input such as prompt wording or temperature can dramatically affect the tone, structure, and direction of the generated output.

Besides, I noticed that the model sometimes produced responses that appeared to be incomplete or abruptly cut off. For example, when using the prompt "In the future, education will", the output ended mid-sentence, suggesting that the model stopped generating before reaching a natural conclusion. This behavior is due to the way GPT-2 handles output length during generation.

Specifically, the parameter max_new_tokens controls the maximum number of new tokens (word pieces) the model is allowed to generate. In my case, the generation stopped after reaching this token limit. While GPT-2 has an end-of-text token (with an ID of 50256), it does not always generate this token naturally unless specifically trained to do so or forced with stopping criteria. As a result, even though the model continued generating up to the limit, it did not include a natural stopping point, leading to a response that felt abruptly cut off.

This experience highlighted an important limitation of open-ended text generation using transformers: without careful tuning of parameters or manual post-processing (e.g., truncating at sentence boundaries), the output may not always be semantically complete. It reinforced the value of understanding how generation parameters like max_new_tokens, temperature, and token limits interact to shape the final result.